File: new/875_koko_eating_bananas_brute_force.py
import math


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # idea: brute force
        # try the speed from 1 until h
        # track the total hours to complete all the piles
        # if we can complete it then we return that speed, as it will be the
        # minimum k as we are calculating from speed = 1
        current_speed = (
            1  # speed of bananas per hour so currently set at 1 banana per hour
        )
        while True:
            total_hours_spent_eating = 0
            for bananas in piles:
                total_hours_spent_eating += math.ceil(bananas / current_speed)
                # bananas / current_speed = number of bananas per (number of bananas per hour) =
                # number of hours, which we add it to tal hours spent eating
                # we use ceil because if a pile has 17 bananas, and our current speed is 5
                # 17 / 5 would be three point something, so we ceil it to 4. so 4 hours spent
                # eating 17 bananas.
            if total_hours_spent_eating <= h:
                return current_speed

            current_speed += 1

        return current_speed
        # time: O(m * n), where m = max(piles), and n = length of piles.
        # space: O(1)


# giving TIME LIMIT EXCEEDED.

File: new/875_koko_eating_bananas_brute_force.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        // idea: brute force
        // from speed = 1 to until the max(piles), try everything and choose the minimum speed
        int speed = 1;
        while (true) {
            long long totalHours = 0;
            for (int bananas: piles) {
                totalHours += (bananas + speed - 1) / speed;
                // in c++ "/" always does integer division and rounds down.
                // so this is the trick we do to ceil up. same as ceil(bananas / speed) in python. 
                // give enough numerator for the denominator to give correct value ceiled up. 
            }

            if (totalHours <= h) {
                return speed;
            }

            speed += 1;
        }

        return speed;
    }
    // time: O(m * n) = where m = max(piles) and n = length of piles
    // space: O(1)
};
